// capacitor-resources/src/generate/platforms/splash/android.js
module.exports = {
    platform: 'android',
    type: 'splash',
    path: 'android/splash/',
    definitions: [{
            name: "drawable-land-hdpi-screen.png",
            width: 800,
            height: 480,
            comment: "land-hdpi"
        }, // 800x480
        {
            name: "drawable-land-ldpi-screen.png",
            width: 320,
            height: 200,
            comment: "land-ldpi"
        }, // 320x200
        {
            name: "drawable-land-mdpi-screen.png",
            width: 480,
            height: 320,
            comment: "land-mdpi"
        }, // 480x320
        {
            name: "drawable-land-xhdpi-screen.png",
            width: 1280,
            height: 720,
            comment: "land-xhdpi"
        }, // 1280x720
        {
            name: "drawable-land-xxhdpi-screen.png",
            width: 1600,
            height: 960,
            comment: "land-xxhdpi"
        }, // 1600x960
        {
            name: "drawable-land-xxxhdpi-screen.png",
            width: 1920,
            height: 1280,
            comment: "land-xxhdpi"
        }, // 1920x1280
        {
            name: "drawable-port-hdpi-screen.png",
            width: 480,
            height: 800,
            comment: "port-hdpi"
        }, // 480x800
        {
            name: "drawable-port-ldpi-screen.png",
            width: 200,
            height: 320,
            comment: "port-ldpi"
        }, // 200x320
        {
            name: "drawable-port-mdpi-screen.png",
            width: 320,
            height: 480,
            comment: "port-mdpi"
        }, // 320x480
        {
            name: "drawable-port-xhdpi-screen.png",
            width: 720,
            height: 1280,
            comment: "port-xhdpi"
        }, // 720x1280
        {
            name: "drawable-port-xxhdpi-screen.png",
            width: 960,
            height: 1600,
            comment: "port-xxhdpi"
        }, // 960x1600
        {
            name: "drawable-port-xxxhdpi-screen.png",
            width: 1280,
            height: 1920,
            comment: "port-xxhdpi"
        } // 1280x1920
    ]
};
